be/src/runtime/sender_queue.cpp

```diff
@@ -0,0 +1,957 @@
+        // Sender send a packet failed, then close the channel.
+        // but closed packet reach first, then the failed packet.
+        // Then meet the assert
+        // we remove the assert
+        // DCHECK_GT(_num_remaining_senders, 0);
+        if (_num_remaining_senders <= 0) {
+            DCHECK(_sender_eos_set.end() != _sender_eos_set.find(be_number));
+            return Status::OK();
+        }
+        // We only need to build chunk meta on first chunk and not use_pass_through
+        // By using pass through, chunks are transmitted in shared memory without ser/deser
+        // So there is no need to build chunk meta.
+        if (_chunk_meta.types.empty() && !use_pass_through) {
+            SCOPED_TIMER(_recvr->_deserialize_chunk_timer);
+            auto& pchunk = request.chunks(0);
+            RETURN_IF_ERROR(_build_chunk_meta(pchunk));
+        }
+    }
+
+    ChunkQueue chunks;
+    size_t total_chunk_bytes = 0;
+    faststring uncompressed_buffer;
+
+    if (use_pass_through) {
+        ChunkUniquePtrVector swap_chunks;
+        std::vector<size_t> swap_bytes;
+        _recvr->_pass_through_context.pull_chunks(request.sender_id(), &swap_chunks, &swap_bytes);
+        DCHECK(swap_chunks.size() == swap_bytes.size());
+        size_t bytes = 0;
+        for (size_t i = 0; i < swap_chunks.size(); i++) {
+            // The sending and receiving of chunks from _pass_through_context may out of order, and
+            // considering the following sequences:
+            // 1. add chunk_1 to _pass_through_context and send request_1
+            // 2. add chunk_2 to _pass_through_context and send request_2
+            // 3. receive request_1 and get both chunk_1 and chunk_2
+            // 4. receive request_2 and get nothing
+            // So one receiving may receive two or more chunks, and we need to use the chunk's driver_sequence
+            // but not the request's driver_sequence
+            ChunkItem item{static_cast<int64_t>(swap_bytes[i]), std::move(swap_chunks[i].first), nullptr};
+            chunks.emplace_back(std::move(item));
+            bytes += swap_bytes[i];
+        }
+        total_chunk_bytes += bytes;
+        COUNTER_UPDATE(_recvr->_bytes_pass_through_counter, total_chunk_bytes);
+
+    } else {
+        for (auto i = 0; i < request.chunks().size(); ++i) {
+            auto& pchunk = request.chunks().Get(i);
+            int64_t chunk_bytes = pchunk.data().size();
+            ChunkUniquePtr chunk = std::make_unique<vectorized::Chunk>();
+            RETURN_IF_ERROR(_deserialize_chunk(pchunk, chunk.get(), &uncompressed_buffer));
+            ChunkItem item{chunk_bytes, std::move(chunk), nullptr};
+            chunks.emplace_back(std::move(item));
+            total_chunk_bytes += chunk_bytes;
+        }
+        COUNTER_UPDATE(_recvr->_bytes_received_counter, total_chunk_bytes);
+    }
+
+    wait_timer.start();
+    {
+        std::lock_guard<Mutex> l(_lock);
```

Comment:
Shall we write a util like `TimedLockGuard` ?

Reply:
And print a warning log if wait a lock for too long ?

Reply:
> Shall we write a util like `TimedLockGuard` ?

I think it is necessary